crypto/asn1_dsa.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * A simple ASN.1 DER encoder/decoder for DSA-Sig-Value and ECDSA-Sig-Value.
+ *
+ * DSA-Sig-Value ::= SEQUENCE {
+ *  r  INTEGER,
+ *  s  INTEGER
+ * }
+ *
+ * ECDSA-Sig-Value ::= SEQUENCE {
+ *  r  INTEGER,
+ *  s  INTEGER
+ * }
+ */
+
+#include <openssl/bn.h>
+#include "internal/asn1_dsa.h"
+
+#define ID_SEQUENCE 0x30
+#define ID_INTEGER 0x02
+
+static size_t calc_der_len(size_t cont_len)
+{
+    size_t llen;
+
+    if (cont_len <= 0x7f)
+        llen = 1;
+    else if (cont_len <= 0xff)
+        llen = 2;
+    else if (cont_len <= 0xffff)
+        llen = 3;
+    else
+        return 0;
+    return 1 + llen + cont_len;
+}
+
+static void enc_len(size_t cont_len, unsigned char **ppout)
+{
+    if (cont_len > 0xff) {
+        *(*ppout)++ = 0x82;
+        *(*ppout)++ = (unsigned char)(cont_len >> 8);
+    } else if (cont_len > 0x7f)
+        *(*ppout)++ = 0x81;
+    *(*ppout)++ = (unsigned char)cont_len;
+}
+
+static size_t calc_int_der_len(const BIGNUM *n)
+{
+    if (BN_is_negative(n))
+        return 0;
+    return calc_der_len(BN_num_bits(n) / 8 + 1);
+}
+
+static void enc_int(const BIGNUM *n, unsigned char **ppout)
+{
+    size_t len = BN_num_bits(n) / 8 + 1;
+
+    *(*ppout)++ = ID_INTEGER;
+    enc_len(len, ppout);
+    BN_bn2binpad(n, *ppout, (int)len);
```

Comment:
BN_bn2binpad() returns -1 if the supplied buffer is too small.
That should never happen since len is calculated to either be equal to the required size or, when BN_num_bits(n) is a multiple of 8, one greater than the required size; in the latter case the extra zero padding on the left is to ensure that the two's complement encoding of the INTEGER content always represents a positive value.

Reply:
I think we should test for such failures anyway. Relying on complex "this should never happen" rules and the internal implementation of BN_bn2binpad is brittle. Code changes over time and suddenly we could find that errors might be returned.

Reply:
Sure. I have added an assert.